packages/@aws-cdk/aws-rds/lib/cluster-engine.ts

```diff
@@ -0,0 +1,318 @@
+import * as ec2 from '@aws-cdk/aws-ec2';
+import * as iam from '@aws-cdk/aws-iam';
+import * as secretsmanager from '@aws-cdk/aws-secretsmanager';
+import * as core from '@aws-cdk/core';
+import { ClusterParameterGroup, IParameterGroup, ParameterGroup } from './parameter-group';
+import { ParameterGroupFamilyMapping } from './private/parameter-group-family-mapping';
+import { compare } from './private/version';
+
+/**
+ * The extra options passed to the {@link IClusterEngine.bindToCluster} method.
+ */
+export interface ClusterEngineBindOptions {
+  /**
+   * The role used for S3 importing.
+   *
+   * @default - none
+   */
+  readonly s3ImportRole?: iam.IRole;
+
+  /**
+   * The role used for S3 exporting.
+   *
+   *  @default - none
+   */
+  readonly s3ExportRole?: iam.IRole;
+
+  /**
+   * The customer-provided ParameterGroup.
+   *
+   * @default - none
+   */
+  readonly parameterGroup?: IParameterGroup;
+}
+
+/**
+ * The type returned from the {@link IClusterEngine.bindToCluster} method.
+ */
+export interface ClusterEngineConfig {
+  /**
+   * The ParameterGroup to use for the cluster.
+   *
+   * @default - no ParameterGroup will be used
+   */
+  readonly parameterGroup?: IParameterGroup;
+
+  /**
+   * The port to use for this cluster,
+   * unless the customer specified the port directly.
+   *
+   * @default - use the default port for clusters (3306)
+   */
+  readonly port?: number;
+}
+
+/**
+ * The interface representing a database cluster (as opposed to instance) engine.
+ */
+export interface IClusterEngine {
+  /** The type of the engine, for example "aurora-mysql". */
+  readonly engineType: string;
+
+  /**
+   * The exact version of a given engine.
+   *
+   * @default - default version for the given engine type
+   */
+  readonly engineVersion?: string;
+
+  /**
+   * The family to use for ParameterGroups using this engine.
+   * This is usually equal to "<engineType><engineMajorVersion>",
+   * but can sometimes be a variation of that.
+   * You can pass this property when creating new ClusterParameterGroup.
+   */
+  readonly parameterGroupFamily: string;
+
+  /** The application used by this engine to perform rotation for a single-user scenario. */
+  readonly singleUserRotationApplication: secretsmanager.SecretRotationApplication;
+
+  /** The application used by this engine to perform rotation for a multi-user scenario. */
+  readonly multiUserRotationApplication: secretsmanager.SecretRotationApplication;
+
+  /** The default EC2 instance type to use for databases in this cluster. */
+  readonly defaultInstanceType: ec2.InstanceType;
```

Comment:
1. Why do we need different instance type defaults per engine? Is this something we already supported in the RDS library?

2. If so, why expose this as a public property? It should be automatically computed in the CDK.

Reply:
1. It's not supported by RDS (`instanceType` is required).
2. It is automatically computed by the CDK....? That's why this property is here. It allows you to make `instanceType` optional in all places that take an `IClusterEngine` or `IInstanceEngine` as a parameter, and allow you to use it in code like:

```ts
new rds.DatabaseInstanceReadReplica(stack, 'ReadReplica', {
  sourceDatabaseInstance: sourceInstance,
  instanceType: rds.DatabaseInstanceEngine.POSTGRES.defaultInstanceType,
  // ...
});
```

Reply:
I'm all for finding the right default and even picking different defaults based on the engine, but exposing them feels off.

I'd rather hide this, and ask the user to specify an instance type for `DatabaseInstanceReadReplica`. 

The other way to do this is to expose `instanceType` as a readonly property of `IDatabaseInstance` and use that

```ts
instanceType: sourceInstance.instanceType,
```

Reply:
When you say

> I'd rather hide this, and ask the user to specify an instance type for `DatabaseInstanceReadReplica`. 

Do you mean you'd rather move this property from `IClusterEngine` to `ClusterEngineConfig` (the type returned from `bindToCluster()`)?

Reply:
Yes, that seems right, especially since we want different defaults per engine.

Reply:
> Yes, that seems right, if we want different defaults per engine.

OK. I'm a little worried about this comment. While I'm not entirely opposed to this change, I want to make sure we're doing this for the right reasons, and this isn't it. So let's discuss further.

Right now, this property is an instance property of `IIClusterEngine`. Which means every engine can have a different default instance type. The fact that all 3 implementations of `IClusterEngine` today have the same default is not due to some limitation that this design imposes, but just the fact that they all accept clusters with this type, so it works out to be a sensible default for all 3.

Does that make sense? Does it change your mind on where should this property be modeled?

Reply:
> Further, given that they are all the same today, why do this now?
> 
> This can be added later when the default for an engine needs to be different. That would likely be a better time to find the correct solution. Until then, use `m5.xlarge` as the default everywhere.

Except they're not the same today. All `IClusterEngine`s use `t3.medium`, yes. But `IInstanceEngine`s use `m5.large`, except SQL Server, which uses `m5.xlarge`.

But let's have it your way. I'll hard-code the default to `t3.medium` in `DatabaseCluster` and to `m5.large` in `DatabaseInstanceSource`.